### PR TITLE
add Enzyme timings to `-ftime-trace` reports

### PR DESCRIPTION
**Pull request by samuelpmish**
This small PR adds a few scoped timers to Enzyme, revealing information about where time is spent in the compilation process.

Before:
Screenshot 2024-05-20 at 4 28 38 PM

After:
Screenshot 2024-05-20 at 4 29 00 PM

### CODE REVIEW

**Comment by samuelpmish** [comment]
oops, I guess I should remove the "test" name in the NVVM bit

**Comment by ZuseZ4** [comment]
Thanks a lot for this!.

**Comment by wsmoses** [comment]
@ZuseZ4 fun fact we got this to also speciy which function is the one taking time.

@samuelpmish can you run clang format before we merge (see CI)?

File: enzyme/Enzyme/ActivityAnalysis.cpp [hunk 1]
```diff
@@ -730,6 +731,9 @@ bool isPossibleFloat(const TypeResults &TR, Value *I, const DataLayout &DL) {
 /// do not propagate adjoints themselves
 bool ActivityAnalyzer::isConstantInstruction(TypeResults const &TR,
                                              Instruction *I) {
+
+  TimeTraceScope timeScope("is constant instruction");
```
Comment (wsmoses) [inline enzyme/Enzyme/ActivityAnalysis.cpp:735]: Can you change this to "ActivityAnalysis (isConstantInstruction)", I->getParent()->getParent()->getName(), I->getName())

File: enzyme/Enzyme/ActivityAnalysis.cpp [hunk 2]
```diff
@@ -730,6 +731,9 @@ bool isPossibleFloat(const TypeResults &TR, Value *I, const DataLayout &DL) {
...
+
```
Comment (wsmoses) [inline enzyme/Enzyme/ActivityAnalysis.cpp:736]: and also add isConstantValue for symmetry

**Comment by ZuseZ4** [comment]
@wsmoses wdym with fun fact? Is there a second motivation for this besides debugging compile time issues that I am not aware of? Or did I missread your message?

File: enzyme/Enzyme/ActivityAnalysis.cpp (hunk 1 above)
Comment (wsmoses) [reply enzyme/Enzyme/ActivityAnalysis.cpp:735]: eh in this case let's just use the last one for scoping reasons

**Comment by wsmoses** [comment]
@ZuseZ4 sorry I should've been more clear. This can also say which function being compiled is the one that takes time (in addition to which function within Enzyme takes time). E.g. runing TA on foo takes all the time, but bar is okay

**Comment by samuelpmish** [comment]
> @samuelpmish can you run clang format before we merge (see CI)?

I ran clang-format in the way described by the contributing document (and it made a few changes), but the format CI job still isn't passing. Does enzyme require a specific version of clang-format?

**Comment by ZuseZ4** [comment]
> > @samuelpmish can you run clang format before we merge (see CI)?
> 
> I ran clang-format in the way described by the contributing document (and it made a few changes), but the format CI job still isn't passing. Does enzyme require a specific version of clang-format?

Iirc yes, I think the formating can change with new versions.
>Run DoozyX/clang-format-lint-action@v0.16.2
  with:
    source: enzyme/Enzyme enzyme/tools/enzyme-tblgen
    exclude: ./CMakeLists.txt enzyme/Enzyme/SCEV
    style: llvm
    clangFormatVersion: 16

**Comment by samuelpmish** [comment]
I just built a copy of llvm 16.06 and tried using the `clang-format` binary from that build on this branch, but it didn't make any changes. Could someone else with an appropriate version of `clang-format` run it on this branch to make it conform to the style guide?

**Comment by ZuseZ4** [comment]
@samuelpmish see (or merge) #1901

**Pull request merged**